src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -3409,6 +3425,53 @@ public void ScanAndProcessReturnValue (MethodBody methodBody)
 				}
 			}
 
+			public void ProcessAttributeDataflow (MethodDefinition method, Collection<CustomAttributeArgument> arguments)
+			{
+				for (int i = 0; i < method.Parameters.Count; i++)
+				{
+					var annotation = _flowAnnotations.GetParameterAnnotation (method, i + 1 /* adjust for `this` */);
+					if (annotation != 0) {
+						ValueNode valueNode = GetValueNodeForCustomAttributeArgument (arguments [i]);
+						if (valueNode != null) {
+							// TODO: There's no way to represent attribute dataflow in current ReflectionPatternContext (and the underlying IReflectionPatterRecorder)
+							ReflectionPatternContext context = new ReflectionPatternContext (_markStep._context, method, method, 0);
+							context.AnalyzingPattern ();
+							RequireDynamicallyAccessedMembers (ref context, annotation, valueNode, method);
+						}
+					}
+				}
+			}
+
+			public void ProcessAttributeDataflow (FieldDefinition field, CustomAttributeArgument value)
+			{
+				var annotation = _flowAnnotations.GetFieldAnnotation (field);
+				Debug.Assert (annotation != 0);
+
+				ValueNode valueNode = GetValueNodeForCustomAttributeArgument (value);
+				if (valueNode != null) {
+					// TODO: There's no way to represent store to a field given current ReflectionPatternContext (and the underlying IReflectionPatterRecorder)
+					var reflectionContext = new ReflectionPatternContext (_markStep._context, field.DeclaringType.Methods [0], field.DeclaringType.Methods [0], 0);
+					reflectionContext.AnalyzingPattern ();
+					RequireDynamicallyAccessedMembers (ref reflectionContext, annotation, valueNode, field);
+				}
+			}
+
+			private ValueNode GetValueNodeForCustomAttributeArgument(CustomAttributeArgument argument)
+			{
+				ValueNode valueNode;
+				if (argument.Type.Name == "Type") {
+					TypeDefinition referencedType = ((TypeReference)argument.Value).Resolve ();
+					valueNode = referencedType == null ? null : new SystemTypeValue (referencedType);
+				} else if (argument.Type.MetadataType == MetadataType.String) {
+					valueNode = new KnownStringValue ((string)argument.Value);
+				} else {
+					Debug.Fail ("We shouldn't have gotten a non-null annotation for this from GetParameterAnnotation.");
```

Comment:
Nit: Recently there's been feedback to now use Debug.Assert/Fail and instead just throw exceptions. Mainly due to test failures being hard to handle (asserts by default popup a dialog and so on) and also because if this got hit in the wild we might get some data back eventually.

Reply:
Automation should ideally add a listener that does something sane (make the unit test fail, or something like that), but I don't think I'll have time to look into that. Made it a throw.